python/hail/table.py

```diff
@@ -1251,8 +1251,7 @@ def types_compatible(left, right):
             left = list(left)
             right = list(right)
             return (types_match(left, right)
-                    or (isinstance(src, MatrixTable)
-                        and len(left) == 1
+                    or (len(left) == 1
                         and len(right) == 1
                         and isinstance(left[0].dtype, tinterval)
                         and left[0].dtype.point_type == right[0].dtype))
```

Comment:
I'm trying to understand this code a bit more so I can competently review this PR.

So we're relaxing the requirement that the source of the expressions are a matrix table. So, previously the following expression was not permitted because the indexer is a table and the index expression is an interval.

```python
t1 = hl.utils.range_table(10)
t2 = hl.utils.range_table(10)
t2 = t2.key_by(span = hl.interval(t2.idx, t2.idx + 5))
t2[t1.idx]
```

but this was OK?

```python
mt = hl.utils.range_matrix_table(10, 10)
t2 = hl.utils.range_table(10)
t2 = t2.key_by(span = hl.interval(t2.idx, t2.idx + 5))
t2[mt.idx]
```

this seems a bit odd to me, but maybe I probably don't understand the implementation strategy.

Reply:
this is correct. This PR makes the first one OK.

Reply:
(also, was only ok if it was `t2[mt.row_idx]`